QUALITY_DECLARATION.md

```diff
@@ -0,0 +1,132 @@
+
+### Public API Declaration
+
+All symbols in the installed headers are considered part of the public API.
+
+All installed headers are in the `include` directory of the package, headers in any other folders are not installed and considered private.
+
+Additionally, there are generated header files which are installed and therefore part of the public API.
+The source templates for these generated headers are in the `resource` folder.
+
+## Change Control Process
+
+TODO fix link
+
+`rcutils` follows the recommended guidelines for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change_control_process).
+
+This includes:
+
+- all changes occur through a pull request
+- all pull request have two peer reviews
+- all pull request must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)
+- all pull request must resolve related documentation changes before merging
+
+## Documentation
+
+### Feature Documentation
+
+TODO fix link
+
+`rcutils` has a [feature list](TODO) and each item in the list links to the corresponding feature documentation.
+There is documentation for all of the features, and new features require documentation before being added.
+
+### Public API Documentation
+
+TODO fix link
+
+`rcutils` has embedded API documentation and it is generated using doxygen and is [hosted](TODO) along side the feature documentation.
+There is documentation for all of the public API, and new additions to the public API require documentation before being added.
+
+### License
+
+The license for `rcutils` is Apache 2.0, and a summary is in each source file, the type is declared in the `package.xml` manifest file, and a full copy of the license is in the `LICENSE` file.
+
+There is an automated test which runs a linter that ensures each file has a license statement.
+
+### Copyright Statements
+
+The copyright holders each provide a statement of copyright in each source code file in `rcutils`.
+
+There is an automated test which runs a linter that ensures each file has at least one copyright statement.
+
+## Testing
+
+### Feature Testing
+
+Each feature in `rcutils` has corresponding tests which simulate typical usage, and they are located in the `test` directory.
+New features are required to have tests before being added.
+
+### Public API Testing
+
+Each part of the public API have tests, and new additions or changes to the public API require tests before being added.
+The tests aim to cover both typical usage and corner cases, but are quantified by contributing to code coverage.
+
+### Coverage
+
+TODO fix link
+
+`rcutils` follows the recommendations for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#coverage), and opts to use branch coverage instead of line coverage.
```

Comment:
The developer guide needs a section on coverage too.

Reply:
[Developer guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#test-coverage) says that the metric to be used when measuring coverage will be branch coverage. However, [REP2004](https://github.com/ros-infrastructure/rep/blob/c2ba03bac23b8f313e47820833fad000dbbe058b/rep-2004.rst) says that metric used in ROS Core packages will be line coverage, aiming for percentage above 95% (in fact talks about branch coverage as something optional).

This comes with two issues:
1. There's an inconsistency to be solved, maybe discussed, for the ROS2 core packages. Will we use branch or line coverage? 
2. There's no percentage coverage required explicitly for packages outside the ROS2 organization, and maybe this has to be changed in the REP2004 to set a desired standard.

Please check this @chapulina

Reply:
The developer guide is out of date still, see the original proposed pr (https://github.com/ros2/ros2_documentation/pull/460) that led to REP-2004's proposal:

> The end goal would be to move most of this into a REP, and then replace this section with the prescriptions on how to achieve the categories for ROS core packages.

Basically, consider the developer guidelines out-of-date until we separate the requirements into the REP and the prescriptions into a (as yet non-existent) pull request against the developer guide lines.

----

That being said, I don't feel strongly about the line versus branch coverage. My thinking is that in some cases it might not be technically feasible to do branch coverage, so I left it as "at least line, but may use branch".

----

With respect to the percentage, that's still being discussed but we will have some recommendation for ROS core packages via the developer guide. What's already there is out-of-date and should be ignored for now.